### PR TITLE
mgr/nfs: Check if pseudo path is absolute path

### PR DESCRIPTION
**Pull request by varshar16**
Fixes: https://tracker.ceph.com/issues/46565
Signed-off-by: Varsha Rao <varao@redhat.com>

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/pybind/mgr/volumes/fs/nfs.py [hunk 1]
```diff
@@ -529,6 +530,9 @@ def create_export(self, fs_name, cluster_id, pseudo_path, read_only, path):
             if not self.check_fs(fs_name):
                 return -errno.ENOENT, "", f"filesystem {fs_name} not found"
 
+            if not isabs(pseudo_path):
+                return -errno.EINVAL, "", "Pseudo path must be an absolute path"
```
[the remark this answers is not in the document]
Comment (varshar16) [reply src/pybind/mgr/volumes/fs/nfs.py]: Done

**Comment by varshar16** [comment]
> Please add a test.

Done

File: src/pybind/mgr/volumes/fs/nfs.py [hunk 2]
```diff
@@ -529,6 +530,9 @@ def create_export(self, fs_name, cluster_id, pseudo_path, read_only, path):
             if not self.check_fs(fs_name):
                 return -errno.ENOENT, "", f"filesystem {fs_name} not found"
 
+            if not isabs(pseudo_path):
+                return -errno.EINVAL, "", f"pseudo path {pseudo_path} is not an absolute path"
```
Comment (batrick) [inline src/pybind/mgr/volumes/fs/nfs.py:534]: Next, please also check that `pseudo_path` is not `/` (also invalid!). Write a test too.

Comment (varshar16) [reply src/pybind/mgr/volumes/fs/nfs.py]: Also done.

**Comment by varshar16** [comment]
jenkins test dashboard backend

**Comment by batrick** [comment]
jenkins test dashboard backend

**Comment by batrick** [comment]
jenkins test dashboard backend

File: src/pybind/mgr/volumes/fs/nfs.py [hunk 3]
```diff
@@ -529,6 +530,10 @@ def create_export(self, fs_name, cluster_id, pseudo_path, read_only, path):
             if not self.check_fs(fs_name):
                 return -errno.ENOENT, "", f"filesystem {fs_name} not found"
 
+            if not isabs(pseudo_path) or pseudo_path == "/":
```
Comment (batrick) [inline src/pybind/mgr/volumes/fs/nfs.py:533]: Consider someone passing "//" as a pseudo_path. We need to collapse the path. Python has a method for this `os.path.normpath`.

Comment (varshar16) [reply src/pybind/mgr/volumes/fs/nfs.py:534]: > Consider someone passing "//" as a pseudo_path. We need to collapse the path. Python has a method for this `os.path.normpath`.

Normpath doesn't work for leading double slash:
https://stackoverflow.com/questions/7816818/why-doesnt-os-normpath-collapse-a-leading-double-slash

**Comment by varshar16** [comment]
Rebased

**Comment by batrick** [comment]
jenkins test make check

**Comment by varshar16** [comment]
Modified the tests as only first test was being executed.
http://qa-proxy.ceph.com/teuthology/varsha-2020-07-23_05:25:38-rados-wip-nfs-add-pseudoroot-path-check-distro-basic-smithi/

**Comment by varshar16** [comment]
Tests Passed:  http://qa-proxy.ceph.com/teuthology/varsha-2020-07-23_10:42:51-rados-wip-nfs-add-pseudoroot-path-check-distro-basic-smithi/

**Comment by varshar16** [comment]
jenkins retest this please

**Comment by varshar16** [comment]
jenkins test dashboard backend

**Comment by varshar16** [comment]
jenkins test dashboard backend

**Comment by batrick** [comment]
jenkins test dashboard backend

**Pull request merged**